mathjax3-ts/input/tex/physics/PhysicsMethods.ts

```diff
@@ -0,0 +1,900 @@
+  macro = ParseUtil.substituteArgs(parser, args, macro);
+  parser.string = ParseUtil.addArgs(parser, macro, parser.string.slice(parser.i));
+  parser.i = 0;
+  if (++parser.macroCount > parser.configuration.options['maxMacros']) {
+    throw new TexError('MaxMacroSub1',
+                        'MathJax maximum macro substitution count exceeded; ' +
+                        'is there a recursive macro call?');
+  }
+};
+
+
+/**
+ * An operator that needs to be parsed (e.g., a gree letter or nabla) and
+ * applied to a possibly fenced expression. By default automatic fences are
+ * parentheses and brakets, with braces being ignored.
+ * @param {TexParser} parser The calling parser.
+ * @param {string} name The macro name.
+ * @param {string} operator The operator expression.
+ * @param {string[]} ...fences List of opening fences that should be
+ *     automatically sized and paired to its corresponding closing fence.
+ */
+PhysicsMethods.OperatorApplication = function(
+  parser: TexParser, name: string, operator: string,
+  ...fences: string[]) {
+  let left = parser.GetNext();
+  let right = pairs[left];
+  if (!right) {
+    parser.Push(new TexParser(operator, parser.stack.env,
+                              parser.configuration).mml());
+    return;
+  }
+  let lfence = '', rfence = '', arg = '';
+  let enlarge = fences.indexOf(left) !== -1;
+  if (left === '{') {
+    arg = parser.GetArgument(name);
+    if (enlarge) {
+      left = '\\{';
+      right = '\\}';
+    }
+  } else {
+      parser.i++;
+      arg = parser.GetUpTo(name, right);
+    }
+  lfence = (enlarge ? '\\left' : '') + left;
+  rfence = (enlarge ? '\\right' : '') + right;
+  let macro = operator + ' ' + lfence + ' ' + arg + ' ' + rfence;
+  parser.Push(new TexParser(macro, parser.stack.env,
+                            parser.configuration).mml());
+};
+
+
+/***********************
+ * Physics package section 2.3
+ * Operators
+ */
+
+// This works better than in the actual physics package.
+/**
+ * Operator expression with automatic fences and optional exponent.
+ * @param {TexParser} parser The calling parser.
+ * @param {string} name The macro name.
+ * @param {boolean=} opt Set to false if no optional exponent is allowed.
+ * @param {string=} id The name of the function if different from name.
+ */
+PhysicsMethods.Expression = function(parser: TexParser, name: string,
+                                     opt: boolean = true, id: string = '') {
+  id = id || name.slice(1);
+  const exp = opt ? parser.GetBrackets(name) : null;
+  let mml = parser.create('token', 'mi', {texClass: TEXCLASS.OP}, id);
+  if (exp) {
+    const sup = new TexParser(exp,
+                              parser.stack.env, parser.configuration).mml();
+    mml = parser.create('node', 'msup', [mml, sup]);
+  }
+  parser.Push(parser.itemFactory.create('fn', mml));
+  if (parser.GetNext() !== '(') {
+    return;
+  }
+  parser.i++;
+  let arg = parser.GetUpTo(name, ')');
```

Comment:
Does this need to take into account nested parens?  I'm not sure what the original package does, but this will not. So `\sin(\cos(x))` would end at the ')' after the 'x', and you would need to do `\sin({\cos(x)})` to get it to work.  Again, this might be in the original LaTeX package as well.

There are other places this same idea may apply.  I just noticed it here, though.

Reply:
I just checked, and the LaTeX physics package *does* process `\sin(\cos(x))` whereas this implementation doesn't, and I also checked the `xparse` documentation, and it does say that delimiters must be balanced (unless you use braces).  I think there may need to be a new function like `GetUpTo()` that skips over parentheses as well as braces in order to get the same results as the original.

Reply:
Nested automatic fences work now. I did it via a specialised stack item. 
The fully balanced fences as in `xparse` does not yet work. In other words an opening fence will greedily take the next corresponding closing fence as the one to stretch.
I have an idea how one can do this. However, this will require a change of treatment of delimiters for packages like `physics` following commands with fenced parameters, which I don't want to touch right now. I will postpone this for 3.1., as it will also be useful for SRE.